### ISSUE TITLE
Question: Dynamically run ReloadConfiguration

### ISSUE THREAD

**Issue by typkrft** [wez/wezterm#1036]
I wasn't sure if this should go under bug or feature, or just in the discussions. 

Scenario: I use (py)wal to generate color schemes. When I generate a color scheme with wal, this isn't picked up as it is only a separate file referenced in the config. CTRL+SHIFT+R reloads the config and applies the new theme. 

I was wonder if there is a way to either run "ReloadConfiguration" dynamically so I could just add it to the wal command, or maybe wezterm should be watching color_scheme_dirs as well.

Regardless thanks for the stellar program. I have looked through some of the documentation but may have missed something. Also sorry for the nebulous categorization of this issue.

**Comment by wez** [comment]
In the latest release, you can use [wezterm.add_to_config_reload(path_to_color_scheme_file)](https://wezfurlong.org/wezterm/config/lua/wezterm/add_to_config_reload_watch_list.html) to have wezterm also watch the color scheme file and reload when it changes

**Comment by typkrft** [comment]
Awesome! Thanks!

**Comment by typkrft** [comment]
Sorry I spoke to soon.  This seems to apply the config in subsequent tabs, but oddly enough ctrl+shift-r no longer seem to update the colors. 

When I generate a new scheme I see this in the logs:
```LOGS
 2021-08-17T00:43:14.296Z INFO  wezterm_mux_server_impl::local > setting up /Users/brandon/.local/share/wezterm/gui-sock-96181
 2021-08-17T00:43:14.425Z INFO  wezterm_gui::termwindow        > OpenGL initialized! Apple M1 4.1 Metal - 71.7.1 is_context_loss_possible=false wezterm version: 20210814-124438-54e29167
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  P00A0607
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  P1CC2B18
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  P2C65934
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  P387787F
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  P4E7B547
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  P5296199
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  P643589F
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  P79cc5d9
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  P86d8997
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  P9CC2B18
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  PaC65934
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  Pb87787F
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  PcE7B547
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  Pd296199
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  Pe43589F
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  Pf9cc5d9
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  Pg9cc5d9
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  Ph0A0607
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  Pl9cc5d9
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  Pj9cc5d9
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  Pk9cc5d9
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  Pm0A0607
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  4 256 #9cc5d9
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  4 257 #0A0607
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  708 #0A0607
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  6 1 bg red brightness 10
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  6 1 bg green brightness 6
 2021-08-17T00:43:21.027Z WARN  wezterm_term::terminalstate::performer > Unhandled OSC  6 1 bg blue brightness 7
```

Config:
```LUA
local wezterm = require 'wezterm';
local home = os.getenv("HOME")
return {
  wezterm.add_to_config_reload_watch_list(home.."/.cache/wal"),
  color_scheme_dirs = {home.."/.cache/wal"},
  color_scheme = "wezterm-wal",
  font = wezterm.font("Iosevka Nerd Font"),
  send_composed_key_when_left_alt_is_pressed=false,
  send_composed_key_when_right_alt_is_pressed=true,
  use_ime = false,
  leader = { key="a", mods="CTRL", timeout_milliseconds=1000 },
}
```
Generated color scheme:
```TOML
        [colors]
        # The default text color
        foreground = "#dccfd5"
        # The default background color
        background = "#11181f"

        # Overrides the cell background color when the current cell is occupied by the
        # cursor and the cursor style is set to Block
        cursor_bg = "#dccfd5"
        # Overrides the text color when the current cell is occupied by the cursor
        cursor_fg = "black"
        # Specifies the border color of the cursor when the cursor style is set to Block,
        # of the color of the vertical or horizontal bar when the cursor style is set to
        # Bar or Underline.
        cursor_border = "#52ad70"

        # the foreground color of selected text
        selection_fg = "black"
        # the background color of selected text
        selection_bg = "#fffacd"

        # The color of the scrollbar "thumb"; the portion that represents the current viewport
        scrollbar_thumb = "#222222"

        # The color of the split lines between panes
        split = "#444444"

        ansi = ["#11181f", "#5D5F94", "#43629C", "#9F609C", "#C071B1", "#5F97A5", "#62A6BD", "#dccfd5"]
        brights = ["#9a9095", "#5D5F94", "#43629C", "#9F609C", "#C071B1", "#5F97A5", "#62A6BD", "#dccfd5"]

```

**Comment by wez** [comment]
Hmm, the `add_to_config_reload_watch_list` should go outside the returned config structure.
I'd suggest something like the below.

Note also: the basename of file you are watching is `wal` but the name of the color scheme in your config is `wezterm-wal` which doesn't match; I think you want `color_scheme = "wal"`.

```lua
local wezterm = require 'wezterm';
local home = wezterm.home_dir;
wezterm.add_to_config_reload_watch_list(home.."/.cache/wal")
return {
  color_scheme_dirs = {home.."/.cache/wal"},
  color_scheme = "wezterm-wal", -- is this name right?
  font = wezterm.font("Iosevka Nerd Font"),
  send_composed_key_when_left_alt_is_pressed=false,
  send_composed_key_when_right_alt_is_pressed=true,
  use_ime = false,
  leader = { key="a", mods="CTRL", timeout_milliseconds=1000 },
}

```

**Comment by typkrft** [comment]
I thought I was watching a directory, my bad. I've moved the that out of the return and updated it to the actual file it's watching. I am still seeing the same behavior though. Subsequent tabs have the correct theme, but the current tab isn't updated and ctrl+shift+r doesn't seem to update it. Each time `wezterm-wal.toml` is generated I see a different iteration of the same warnings. So it seems to recognize the file changing. 

```LUA
local wezterm = require 'wezterm';
local home = wezterm.home_dir;
wezterm.add_to_config_reload_watch_list(home.."/.cache/wal/wezterm-wal.toml")
return {
  -- Wal Colorscheme NOTE: you could definte multiple colorschemes in this file 
  color_scheme_dirs = {home.."/.cache/wal"},
  color_scheme = "wezterm-wal",
  font = wezterm.font("Iosevka Nerd Font"),
  send_composed_key_when_left_alt_is_pressed=false,
  send_composed_key_when_right_alt_is_pressed=true,
  use_ime = false,
  leader = { key="a", mods="CTRL", timeout_milliseconds=1000 },
  -- keys = {
    -- Figure out how to move left and right in terminal with cmd and opt
  -- }
  -- wezterm.add_to_config_reload_watch_list(home.."/.cache/wal"),
}

}
```

**Comment by typkrft** [comment]
Get ready for the weirdest thing I've seen in a while. This consistently does not work until I open the 6th tab. After I open 6 tabs it works automatically as expected but only only on the 6th tab and up. I'll make a video tomorrow.

**Comment by wez** [comment]
I used this config:

```lua
local wezterm = require 'wezterm';
wezterm.add_to_config_reload_watch_list("/tmp/issue1036/colors/wezterm-wal.toml");
return {
  color_scheme_dirs = {"/tmp/issue1036/colors"},
  color_scheme = "wezterm-wal",
}
```

and with your color scheme file saved as `/tmp/issue1036/colors/wezterm-wal.toml`, making changes to the toml file are immediately picked up in wezterm.

I don't know much about pywal, but assuming that you're talking about https://github.com/dylanaraps/pywal, it looks like pywal unconditionally emits a bunch of escape sequences for different terminals that might also impact the palette; the logs you pasted show some non-standard sequences that might be specified to the linux terminal or some versions of iterm.  Those sequences are indiscriminately sent to all ttys regardless of what might be on the other end (which is honestly a bit scary! https://github.com/dylanaraps/pywal/blob/master/pywal/sequences.py#L76).  I don't know if it is sending sequences that are supported by wezterm (I did't read that deep).

When escape sequences are used to amend the palette, those escape sequences take precedence over the color scheme configuration file, so it's possible that pywal may be sabotaging your efforts at setting up the color scheme if it is updating some but not all of the palette through escape sequences, or otherwise getting in a confused state.
